src/validation.cpp

```diff
@@ -380,10 +380,8 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
     auto it = disconnectpool.queuedTx.get<insertion_order>().rbegin();
     while (it != disconnectpool.queuedTx.get<insertion_order>().rend()) {
         // ignore validation errors in resurrected transactions
-        TxValidationState stateDummy;
         if (!fAddToMempool || (*it)->IsCoinBase() ||
-            !AcceptToMemoryPool(mempool, stateDummy, *it,
-                                nullptr /* plTxnReplaced */, true /* bypass_limits */)) {
+            !AcceptToMemoryPool(mempool, *it, true /* bypass_limits */).m_accepted.value()) {
```

Comment:
abc6ff1

I don't think that's a good idea to encumber code path like `UpdateMempoolForReorg` with `std::optional`. If this `nullopt` will throw an exception. And we do have the failure/unfinished constructor allowing `m_accepted` to be initialized to nullopt, even if AFAICT such constructor is never called with `finished=bad`for now ? 

I know there is a discussion about `std::optional` usage [here](https://github.com/bitcoin/bitcoin/pull/20833#discussion_r557185885) but could we restrain its usage to only `m_base_fee`/`m_replaced_transaction` ? They would be `nullopt` if `m_accepted=false`.

Reply:
Check now? I think it's better with enum

Reply:
Do we really need an enum and can't we rely only on a boolean ? Maybe you can point me to a branch how you're using UNFINISHED because we don't use it with this PR ?

Reply:
Haven't published the branch yet but the idea is to return a `std::vector<MempoolAcceptResult>` from `ProcessNewPackage`, quit early when a tx fails, and set the not-fully-validated txns in the package to `UNFINISHED`. See [comment](https://github.com/bitcoin/bitcoin/pull/20833#discussion_r567952639)?

Reply:
I think `UNFINISHED`/not_fully_validated to mark package partial failure doesn't bring further value compared to unvalid. Do you plan to consume this `UNFINISHED` in a special way ? Otherwise we can just extend `TxValidationResult` with a `PACKAGE_PARENT_FAILED` instead of yet-another-state variable.

Reply:
I think in the future with Package Relay, we may want to punish nodes differently for invalidity in packages vs invalidity in transactions. Similarly, it may affect how we cache failed transactions. Obviously this isn't set in stone, but I think it would be better no to put package-specific validation info in `TxValidationState` - what do you think?

Reply:
I don't see a straightforward reason to punish faulty pacakge-relay peers from regular tx-relay ones, at least at the mempool level. If we don't have a motivation for `UNFINISHED`, let's remove it for now and defer its introduction when we actually hit the case ? Would be easier to argue at that point.